File: router/src/hamming74.py
from dataclasses import dataclass


DATA_BIT_COUNT = 4
CODEWORD_BIT_COUNT = 7


@dataclass(frozen=True)
class RecoveredDataBits:
    data_bits: str
    corrected_block_count: int


def encode_data_bits(data_bits: str) -> str:
    validate_binary_bits(data_bits, "Los bits de datos")

    if len(data_bits) % DATA_BIT_COUNT != 0:
        raise ValueError(
            "Los bits de datos deben tener una cantidad múltiplo de 4."
        )

    encoded_blocks = []

    for start_index in range(0, len(data_bits), DATA_BIT_COUNT):
        data_block = data_bits[start_index:start_index + DATA_BIT_COUNT]
        encoded_blocks.append(encode_data_block(data_block))

    return "".join(encoded_blocks)


def recover_data_bits(codeword_bits: str) -> RecoveredDataBits:
    validate_binary_bits(codeword_bits, "Los bits de la trama")

    if len(codeword_bits) % CODEWORD_BIT_COUNT != 0:
        raise ValueError(
            "Los bits de la trama deben tener una cantidad múltiplo de 7."
        )

    recovered_data_blocks = []
    corrected_block_count = 0

    for start_index in range(0, len(codeword_bits), CODEWORD_BIT_COUNT):
        codeword_block = codeword_bits[
            start_index:start_index + CODEWORD_BIT_COUNT
        ]
        recovered_data_block, block_was_corrected = recover_data_block(
            codeword_block
        )
        recovered_data_blocks.append(recovered_data_block)

        if block_was_corrected:
            corrected_block_count += 1

    return RecoveredDataBits(
        data_bits="".join(recovered_data_blocks),
        corrected_block_count=corrected_block_count,
    )


def encode_data_block(data_block: str) -> str:
    validate_binary_bits(data_block, "El bloque de datos")

    if len(data_block) != DATA_BIT_COUNT:
        raise ValueError("Un bloque de datos Hamming debe tener exactamente 4 bits.")

    codeword_by_position = [None, 0, 0, 0, 0, 0, 0, 0]
    codeword_by_position[3] = int(data_block[0])
    codeword_by_position[5] = int(data_block[1])
    codeword_by_position[6] = int(data_block[2])
    codeword_by_position[7] = int(data_block[3])

    codeword_by_position[1] = calculate_even_parity(
        codeword_by_position,
        [3, 5, 7],
    )
    codeword_by_position[2] = calculate_even_parity(
        codeword_by_position,
        [3, 6, 7],
    )
    codeword_by_position[4] = calculate_even_parity(
        codeword_by_position,
        [5, 6, 7],
    )

    return "".join(
        str(codeword_by_position[position])
        for position in range(1, CODEWORD_BIT_COUNT + 1)
    )


def recover_data_block(codeword_block: str) -> tuple[str, bool]:
    validate_binary_bits(codeword_block, "El bloque Hamming")

    if len(codeword_block) != CODEWORD_BIT_COUNT:
        raise ValueError("Un bloque Hamming debe tener exactamente 7 bits.")

    codeword_by_position = [None] + [int(bit) for bit in codeword_block]
    error_position = calculate_error_position(codeword_by_position)
    block_was_corrected = error_position != 0

    if block_was_corrected:
        codeword_by_position[error_position] = (
            1 - codeword_by_position[error_position]
        )

    recovered_data_bits = "".join(
        str(codeword_by_position[position])
        for position in [3, 5, 6, 7]
    )
    return recovered_data_bits, block_was_corrected


def calculate_even_parity(
    codeword_by_position: list[int | None],
    data_positions: list[int],
) -> int:
    parity = 0

    for position in data_positions:
        parity ^= codeword_by_position[position]

    return parity


def calculate_error_position(codeword_by_position: list[int | None]) -> int:
    parity_check_for_position_1 = calculate_parity_check(
        codeword_by_position,
        [1, 3, 5, 7],
    )
    parity_check_for_position_2 = calculate_parity_check(
        codeword_by_position,
        [2, 3, 6, 7],
    )
    parity_check_for_position_4 = calculate_parity_check(
        codeword_by_position,
        [4, 5, 6, 7],
    )

    return (
        parity_check_for_position_1
        + (parity_check_for_position_2 * 2)
        + (parity_check_for_position_4 * 4)
    )


def calculate_parity_check(
    codeword_by_position: list[int | None],
    checked_positions: list[int],
) -> int:
    parity_check = 0

    for position in checked_positions:
        parity_check ^= codeword_by_position[position]

    return parity_check


def validate_binary_bits(bits: str, description: str) -> None:
    if not isinstance(bits, str):
        raise ValueError(f"{description} deben ser texto.")

    for bit in bits:
        if bit not in {"0", "1"}:
            raise ValueError(f"{description} solo pueden contener 0 y 1.")

File: servidor-java/src/link/Hamming74.java
package link;


public final class Hamming74 {
    private static final int DATA_BIT_COUNT = 4;
    private static final int CODEWORD_BIT_COUNT = 7;

    private Hamming74() {
    }

    public static String encodeDataBits(String dataBits) {
        validateBinaryBits(dataBits, "Los bits de datos");

        if (dataBits.length() % DATA_BIT_COUNT != 0) {
            throw new IllegalArgumentException(
                "Los bits de datos deben tener una cantidad múltiplo de 4."
            );
        }

        StringBuilder encodedBits = new StringBuilder();

        for (
            int startIndex = 0;
            startIndex < dataBits.length();
            startIndex += DATA_BIT_COUNT
        ) {
            String dataBlock = dataBits.substring(
                startIndex,
                startIndex + DATA_BIT_COUNT
            );
            encodedBits.append(encodeDataBlock(dataBlock));
        }

        return encodedBits.toString();
    }

    public static RecoveredDataBits recoverDataBits(String codewordBits) {
        validateBinaryBits(codewordBits, "Los bits de la trama");

        if (codewordBits.length() % CODEWORD_BIT_COUNT != 0) {
            throw new IllegalArgumentException(
                "Los bits de la trama deben tener una cantidad múltiplo de 7."
            );
        }

        StringBuilder recoveredDataBits = new StringBuilder();
        int correctedBlockCount = 0;

        for (
            int startIndex = 0;
            startIndex < codewordBits.length();
            startIndex += CODEWORD_BIT_COUNT
        ) {
            String codewordBlock = codewordBits.substring(
                startIndex,
                startIndex + CODEWORD_BIT_COUNT
            );
            RecoveredDataBlock recoveredDataBlock = recoverDataBlock(
                codewordBlock
            );
            recoveredDataBits.append(recoveredDataBlock.dataBits);

            if (recoveredDataBlock.wasCorrected) {
                correctedBlockCount++;
            }
        }

        return new RecoveredDataBits(
            recoveredDataBits.toString(),
            correctedBlockCount
        );
    }

    private static String encodeDataBlock(String dataBlock) {
        int[] codewordByPosition = new int[CODEWORD_BIT_COUNT + 1];
        codewordByPosition[3] = dataBlock.charAt(0) - '0';
        codewordByPosition[5] = dataBlock.charAt(1) - '0';
        codewordByPosition[6] = dataBlock.charAt(2) - '0';
        codewordByPosition[7] = dataBlock.charAt(3) - '0';

        codewordByPosition[1] = calculateEvenParity(
            codewordByPosition,
            new int[] {3, 5, 7}
        );
        codewordByPosition[2] = calculateEvenParity(
            codewordByPosition,
            new int[] {3, 6, 7}
        );
        codewordByPosition[4] = calculateEvenParity(
            codewordByPosition,
            new int[] {5, 6, 7}
        );

        StringBuilder codewordBits = new StringBuilder();

        for (int position = 1; position <= CODEWORD_BIT_COUNT; position++) {
            codewordBits.append(codewordByPosition[position]);
        }

        return codewordBits.toString();
    }

    private static RecoveredDataBlock recoverDataBlock(String codewordBlock) {
        int[] codewordByPosition = new int[CODEWORD_BIT_COUNT + 1];

        for (int position = 1; position <= CODEWORD_BIT_COUNT; position++) {
            codewordByPosition[position] = codewordBlock.charAt(position - 1) - '0';
        }

        int errorPosition = calculateErrorPosition(codewordByPosition);
        boolean wasCorrected = errorPosition != 0;

        if (wasCorrected) {
            codewordByPosition[errorPosition] = 1 - codewordByPosition[errorPosition];
        }

        String dataBits = ""
            + codewordByPosition[3]
            + codewordByPosition[5]
            + codewordByPosition[6]
            + codewordByPosition[7];
        return new RecoveredDataBlock(dataBits, wasCorrected);
    }

    private static int calculateEvenParity(
        int[] codewordByPosition,
        int[] dataPositions
    ) {
        int parity = 0;

        for (int position : dataPositions) {
            parity ^= codewordByPosition[position];
        }

        return parity;
    }

    private static int calculateErrorPosition(int[] codewordByPosition) {
        int parityCheckForPosition1 = calculateParityCheck(
            codewordByPosition,
            new int[] {1, 3, 5, 7}
        );
        int parityCheckForPosition2 = calculateParityCheck(
            codewordByPosition,
            new int[] {2, 3, 6, 7}
        );
        int parityCheckForPosition4 = calculateParityCheck(
            codewordByPosition,
            new int[] {4, 5, 6, 7}
        );

        return parityCheckForPosition1
            + (parityCheckForPosition2 * 2)
            + (parityCheckForPosition4 * 4);
    }

    private static int calculateParityCheck(
        int[] codewordByPosition,
        int[] checkedPositions
    ) {
        int parityCheck = 0;

        for (int position : checkedPositions) {
            parityCheck ^= codewordByPosition[position];
        }

        return parityCheck;
    }

    private static void validateBinaryBits(String bits, String description) {
        if (bits == null) {
            throw new IllegalArgumentException(description + " deben ser texto.");
        }

        for (int index = 0; index < bits.length(); index++) {
            char bit = bits.charAt(index);

            if (bit != '0' && bit != '1') {
                throw new IllegalArgumentException(
                    description + " solo pueden contener 0 y 1."
                );
            }
        }
    }

    public static final class RecoveredDataBits {
        private final String dataBits;
        private final int correctedBlockCount;

        private RecoveredDataBits(String dataBits, int correctedBlockCount) {
            this.dataBits = dataBits;
            this.correctedBlockCount = correctedBlockCount;
        }

        public String getDataBits() {
            return dataBits;
        }

        public int getCorrectedBlockCount() {
            return correctedBlockCount;
        }
    }

    private static final class RecoveredDataBlock {
        private final String dataBits;
        private final boolean wasCorrected;

        private RecoveredDataBlock(String dataBits, boolean wasCorrected) {
            this.dataBits = dataBits;
            this.wasCorrected = wasCorrected;
        }
    }
}
